### PR TITLE
PF-3004: Fix stairctl so it runs again, other updates

### PR DESCRIPTION
**Pull request by pshapiro4broad**
This PR updates the configuration of `stairctl` so it works again.

While updating it, I noticed that it makes heavy use of `System.out`, which isn't the recommended approach for Spring Shell apps. I updated it to use `Terminal` instead.

While testing, I've used this to look at TDR flight logs in dev and locally, and it seems like it might be useful. It may be useful for other Terra apps that use stairway, for developers, or a tool for support, so they can look at flight failures in more detail without needing to search production logs.

Here's an example session using `stairctl`:

```
❯ java -jar stairctl/build/libs/stairctl-1.1.9-SNAPSHOT.jar
StairCtl  (v1.1.9-SNAPSHOT)

[main] INFO bio.terra.stairctl.Application - Starting Application v1.1.9-SNAPSHOT using Java 17.0.4.1 with PID 68497 (/Users/pshapiro/source/stairway/stairctl/build/libs/stairctl-1.1.9-SNAPSHOT.jar started by pshapiro in /Users/pshapiro/source/stairway)
[main] INFO bio.terra.stairctl.Application - No active profile set, falling back to 1 default profile: "default"
[main] INFO bio.terra.stairctl.Application - Started Application in 2.031 seconds (process running for 2.434)
stairctl> connect
Connected to Stairway
Connected to Stairway on database stairwaylib
stairctl> show connection
Stairway Connection:
  dbname  : stairwaylib
  username: stairwayuser
  host    : 127.0.0.1
  port    : 5432
stairctl> list flights

Offset FlightId                             Class                                Submitted                   Completed                   Status       StairwayId                    
------ ------------------------------------ ------------------------------------ --------------------------- --------------------------- ------------ ------------------------------
     0 undoTest                             b.t.s.flights.TestFlightRecoveryUndo 2024-07-12T15:36:39.407627Z 2024-07-12T15:36:44.511316Z ERROR                                      

stairctl> get flight --flightId undoTest

Flight: undoTest
  class     : bio.terra.stairway.flights.TestFlightRecoveryUndo
  submitted : 2024-07-12T15:36:39.407627Z
  completed : 2024-07-12T15:36:44.511316Z
  duration  : 0:00:05.103
  status    : ERROR
  exception : java.lang.RuntimeException;TestStepTriggerUndo
  stairwayId: 
stairctl> 
```

### CODE REVIEW

File: stairctl/src/main/resources/application.yml [hunk 1]
```diff
@@ -7,3 +7,10 @@ stairctl:
   host: ${STAIRCTL_HOST:127.0.0.1}
   port: ${STAIRCTL_PORT:5432}
   dbname: ${STAIRCTL_DBNAME:stairwaylib}
+
+spring:
+  shell:
+    interactive:
```
Comment (pshapiro4broad) [inline stairctl/src/main/resources/application.yml:13]: Spring shell 3.3 disables the interactive shell by default, so we need to set this property to enable it.

https://github.com/spring-projects/spring-shell/wiki/Spring-Shell-3.3-Release-Notes

File: stairctl/build.gradle [hunk 2]
```diff
@@ -19,7 +19,11 @@ dependencies {
...
     // Spring
-    implementation group: 'org.springframework.boot', name: 'spring-boot-starter'
+    implementation group: 'org.springframework.boot', name: 'spring-boot'
     // Spring Boot and Spring Shell versions should stay in sync
-    implementation group: 'org.springframework.shell', name: 'spring-shell-starter', version: '3.3.1'
+    implementation('org.springframework.shell:spring-shell-starter:3.3.1') {
+        exclude module: 'spring-boot-starter-logging'
+    }
+
+    implementation 'org.slf4j:slf4j-simple:1.7.36'
```
Comment (pshapiro4broad) [inline stairctl/build.gradle]: This isn't necessary but it cleans up the log configuration errors you get on startup

File: stairctl/src/main/java/bio/terra/stairctl/commands/ConnectCommands.java [hunk 3]
```diff
@@ -31,47 +33,41 @@ public void connect(
               defaultValue = ShellOption.NULL)
           String dbname,
       @ShellOption(
-              value = {"-h", "--host"},
+              value = {"-H", "--host"},
```
Comment (pshapiro4broad) [inline stairctl/src/main/java/bio/terra/stairctl/commands/ConnectCommands.java:36]: This was required as `-h` conflicts with `help`, which is a global option. Spring shell didn't report an error but it did cause some odd problems when trying to use other options for this command.

File: stairctl/build.gradle (hunk 2 above)
Comment (okotsopoulos) [reply stairctl/build.gradle]: Fair enough, would you mind adding a comment to that effect on the exclusion?

File: stairctl/src/main/java/bio/terra/stairctl/ConnectParams.java [hunk 4]
```diff
@@ -1,71 +1,23 @@
...
-
-  public String getDbname() {
-    return dbname;
-  }
-
-  public ConnectParams dbname(String dbname) {
-    this.dbname = dbname;
-    return this;
-  }
+public record ConnectParams(
```
Comment (okotsopoulos) [inline stairctl/src/main/java/bio/terra/stairctl/ConnectParams.java:7]: Nice, love a config-to-record refactor.

File: stairctl/src/main/java/bio/terra/stairctl/ConnectParams.java [hunk 5]
```diff
@@ -1,71 +1,23 @@
...
   // Fill in any null parameters with those from the provided defaults
-  public void applyDefaults(ConnectParams defaults) {
-    username = Optional.ofNullable(username).orElse(defaults.getUsername());
-    password = Optional.ofNullable(password).orElse(defaults.getPassword());
-    host = Optional.ofNullable(host).orElse(defaults.getHost());
-    port = Optional.ofNullable(port).orElse(defaults.getPort());
-    dbname = Optional.ofNullable(dbname).orElse(defaults.getDbname());
+  public ConnectParams withDefaults(ConnectParams defaults) {
+    return new ConnectParams(
+        Objects.requireNonNullElseGet(username, defaults::username),
```
Comment (okotsopoulos) [inline stairctl/src/main/java/bio/terra/stairctl/ConnectParams.java:17]: TIL about this Objects helper method.

**Review by okotsopoulos — APPROVED** [review]
Thank you for noodling on this!  Would you mind also looking at the stairctl readme to see if any updates are needed in light of these changes?

I think that the ticket number also needs to be associated with this PR - https://broadworkbench.atlassian.net/browse/PF-3004

File: buildSrc/src/main/groovy/stairway.java-conventions.gradle [hunk 6]
```diff
@@ -14,7 +13,7 @@ repositories {
 
 dependencies {
     // For logging
-    implementation group: 'org.slf4j', name: 'slf4j-api', version: '1.7.36'
+    implementation group: 'org.slf4j', name: 'slf4j-api', version: '2.0.13'
```
Comment (pshapiro4broad) [inline buildSrc/src/main/groovy/stairway.java-conventions.gradle:16]: I was able to remove the stairctl subproject dependency exclusion by updating the slf4j version to the current release. This required changing the logback config files to xml, as groovy was deprecated a while ago due to security issues.

File: stairctl/src/main/resources/application.yml [hunk 7]
```diff
@@ -7,3 +7,15 @@ stairctl:
...
+      enabled: true
+    script:
+      enabled: true
+
+# Disable logging
```
Comment (pshapiro4broad) [inline stairctl/src/main/resources/application.yml:18]: While looking the logging dependencies, I thought that, as a command line program, stairctl shouldn't log at all, so I added a config to `properties` to disable logging.

**Review by fboulnois — APPROVED** [review]
👍

**Pull request merged**